// vue-amazing-ui/es/descriptions/descriptions/Descriptions.vue2.js
import { defineComponent as I, ref as i, computed as z, watch as F, onMounted as G, nextTick as k, createElementBlock as s, openBlock as o, normalizeClass as V, createCommentVNode as H, withDirectives as J, createElementVNode as d, renderSlot as w, createTextVNode as M, toDisplayString as O, Fragment as f, renderList as h, vShow as q } from "vue";
import { useSlotsExist as K, useEventListener as P, useMutationObserver as Q } from "../../utils/index.js";
const U = {
  key: 0,
  class: "descriptions-header"
}, X = { class: "descriptions-title" }, Y = { class: "descriptions-extra" }, Z = { key: 0 }, ee = ["colspan"], te = { key: 1 }, re = { key: 0 }, le = ["colspan"], se = ["colspan"], oe = { key: 1 }, ie = /* @__PURE__ */ I({
  __name: "Descriptions",
  props: {
    title: { default: void 0 },
    extra: { default: void 0 },
    bordered: { type: Boolean, default: !1 },
    vertical: { type: Boolean, default: !1 },
    size: { default: "default" },
    column: { default: () => ({ xs: 1, sm: 2, md: 3 }) },
    labelStyle: { default: () => ({}) },
    contentStyle: { default: () => ({}) }
  },
  setup(W) {
    const r = W, S = i(), C = i(!0), m = i(!0), p = i(), B = i(), E = i(), R = i(), g = i(), x = i(), D = i(), c = i([]), v = i(window.innerWidth), L = K(["title", "extra"]), A = z(() => L.title || L.extra || r.title || r.extra), y = z(() => typeof r.column == "object" ? v.value >= 1600 && r.column.xxl !== void 0 ? r.column.xxl : v.value >= 1200 && r.column.xl !== void 0 ? r.column.xl : v.value >= 992 && r.column.lg !== void 0 ? r.column.lg : v.value >= 768 && r.column.md !== void 0 ? r.column.md : v.value >= 576 && r.column.sm !== void 0 ? r.column.sm : v.value < 576 && r.column.xs !== void 0 ? r.column.xs : 1 : r.column);
    F(
      () => [r.bordered, r.vertical, y.value, r.labelStyle, r.contentStyle],
      () => {
        m.value || (m.value = !0), N();
      },
      {
        deep: !0
      }
    ), P(window, "resize", T), Q(
      S,
      (t) => {
        m.value || (m.value = !0, t.some((e) => e.type === "childList") && N());
      },
      { subtree: !0, childList: !0, attributes: !0 }
    ), G(() => {
      $();
    });
    function T() {
      v.value = window.innerWidth;
    }
    async function N() {
      C.value = !C.value, await k(), $();
    }
    function b(t) {
      return t.reduce((l, e) => l + e.span, 0);
    }
    async function $() {
      if (p.value = Array.from(S.value.children).filter((t) => t.className === (r.bordered ? "descriptions-item-bordered" : "descriptions-item")), c.value.length && (c.value = [], await k()), p.value && p.value.length) {
        const t = p.value.length;
        let l = [];
        for (let e = 0; e < t; e++) {
          const n = {
            span: Math.min(p.value[e].dataset.span ?? 1, y.value),
            element: p.value[e]
          };
          b(l) < y.value ? (n.span = Math.min(n.span, y.value - b(l)), l.push(n)) : (c.value.push(l), l = [n]);
        }
        if (!r.vertical && !p.value[t - 1].dataset.span && b(l) < y.value) {
          const e = l.length;
          l[e - 1].span = l[e - 1].span + y.value - b(l);
        }
        c.value.push(l), await k(), j();
      } else
        m.value = !1;
    }
    async function j() {
      r.bordered ? c.value.forEach((t, l) => {
        t.forEach((e) => {
          const n = Array.from(e.element.children), a = n[0];
          _(a, r.labelStyle);
          const u = n[1];
          _(u, r.contentStyle), r.vertical ? (a.colSpan = e.span, u.colSpan = e.span, x.value[l].appendChild(a), D.value[l].appendChild(u)) : (a.colSpan = 1, u.colSpan = e.span * 2 - 1, g.value[l].appendChild(a), g.value[l].appendChild(u));
        });
      }) : p.value.forEach((t, l) => {
        const e = Array.from(t.children), n = e[0];
        _(n, r.labelStyle);
        const a = e[1];
        _(a, r.contentStyle), r.vertical ? (E.value[l].appendChild(t.firstChild), R.value[l].appendChild(t.lastChild)) : B.value[l].appendChild(t);
      }), await k(), m.value = !1;
    }
    function _(t, l) {
      JSON.stringify(l) !== "{}" && Object.keys(l).forEach((e) => {
        t.style[e] || (t.style[e] = l[e]);
      });
    }
    return (t, l) => (o(), s("div", {
      class: V(["descriptions-wrap", `descriptions-${t.size}`])
    }, [
      A.value ? (o(), s("div", U, [
        d("div", X, [
          w(t.$slots, "title", {}, () => [
            M(O(t.title), 1)
          ], !0)
        ]),
        d("div", Y, [
          w(t.$slots, "extra", {}, () => [
            M(O(t.extra), 1)
          ], !0)
        ])
      ])) : H("", !0),
      t.vertical ? (o(), s("div", {
        key: 2,
        class: V(["descriptions-view", { "descriptions-bordered": t.bordered }])
      }, [
        d("table", null, [
          t.bordered ? (o(), s("tbody", oe, [
            (o(!0), s(f, null, h(c.value.length, (e) => (o(), s(f, { key: e }, [
              d("tr", {
                ref_for: !0,
                ref_key: "thVerticalBorderedRows",
                ref: x,
                class: "descriptions-bordered-tr"
              }, null, 512),
              d("tr", {
                ref_for: !0,
                ref_key: "tdVerticalBorderedRows",
                ref: D,
                class: "descriptions-bordered-tr"
              }, null, 512)
            ], 64))), 128))
          ])) : (o(), s("tbody", re, [
            (o(!0), s(f, null, h(c.value, (e, n) => (o(), s(f, { key: n }, [
              d("tr", null, [
                (o(!0), s(f, null, h(e, (a, u) => (o(), s("th", {
                  class: "descriptions-item-th",
                  colspan: a.span,
                  key: u
                }, [
                  d("div", {
                    ref_for: !0,
                    ref_key: "thVerticalCols",
                    ref: E,
                    class: "descriptions-item"
                  }, null, 512)
                ], 8, le))), 128))
              ]),
              d("tr", null, [
                (o(!0), s(f, null, h(e, (a, u) => (o(), s("td", {
                  class: "descriptions-item-td",
                  colspan: a.span,
                  key: u
                }, [
                  d("div", {
                    ref_for: !0,
                    ref_key: "tdVerticalCols",
                    ref: R,
                    class: "descriptions-item"
                  }, null, 512)
                ], 8, se))), 128))
              ])
            ], 64))), 128))
          ]))
        ])
      ], 2)) : (o(), s("div", {
        key: 1,
        class: V(["descriptions-view", { "descriptions-bordered": t.bordered }])
      }, [
        d("table", null, [
          t.bordered ? (o(), s("tbody", te, [
            (o(!0), s(f, null, h(c.value.length, (e) => (o(), s("tr", {
              ref_for: !0,
              ref_key: "trBorderedRows",
              ref: g,
              class: "descriptions-bordered-tr",
              key: e
            }))), 128))
          ])) : (o(), s("tbody", Z, [
            (o(!0), s(f, null, h(c.value, (e, n) => (o(), s("tr", { key: n }, [
              (o(!0), s(f, null, h(e, (a, u) => (o(), s("td", {
                ref_for: !0,
                ref_key: "tdCols",
                ref: B,
                class: "descriptions-item-td",
                colspan: a.span,
                key: u
              }, null, 8, ee))), 128))
            ]))), 128))
          ]))
        ])
      ], 2)),
      J(d("div", {
        ref_key: "defaultSlotsRef",
        ref: S
      }, [
        C.value ? w(t.$slots, "default", { key: 0 }, void 0, !0) : w(t.$slots, "default", { key: 1 }, void 0, !0)
      ], 512), [
        [q, !1]
      ])
    ], 2));
  }
});
export {
  ie as default
};
