// vue-amazing-ui/es/countdown/Countdown.vue2.js
import { defineComponent as b, ref as k, computed as S, watch as T, onMounted as z, createElementBlock as u, openBlock as s, createCommentVNode as m, createElementVNode as I, normalizeStyle as F, renderSlot as y, createTextVNode as M, toDisplayString as h, Fragment as A } from "vue";
import { useSlotsExist as P } from "../utils/index.js";
const Z = { class: "countdown-wrap" }, j = { class: "countdown-time" }, G = {
  key: 0,
  class: "time-prefix"
}, J = {
  key: 0,
  class: "time-suffix"
}, O = /* @__PURE__ */ b({
  __name: "Countdown",
  props: {
    title: { default: void 0 },
    titleStyle: { default: () => ({}) },
    prefix: { default: void 0 },
    suffix: { default: void 0 },
    finish: { default: void 0 },
    future: { type: Boolean, default: !0 },
    format: { default: "HH:mm:ss" },
    value: { default: 0 },
    valueStyle: { default: () => ({}) },
    active: { type: Boolean, default: !0 }
  },
  emits: ["finish"],
  setup(B, { expose: E, emit: N }) {
    const l = B, r = k(0), a = k(0), o = k(null), $ = N, D = P(["title", "prefix", "suffix"]), q = S(() => D.title || l.title), Y = S(() => D.prefix || l.prefix), C = S(() => D.suffix || l.suffix), n = S(() => ({
      showMillisecond: l.format.includes("SSS"),
      showYear: l.format.includes("Y"),
      showMonth: l.format.includes("M"),
      showDay: l.format.includes("D"),
      showHour: l.format.includes("H"),
      showMinute: l.format.includes("m"),
      showSecond: l.format.includes("s")
    }));
    T(
      () => l.active,
      (e) => {
        l.future || (e ? (r.value = a.value + Date.now(), o.value = requestAnimationFrame(g)) : (o.value && cancelAnimationFrame(o.value), o.value = null));
      }
    ), T(
      () => [l.value, l.future],
      () => {
        x();
      },
      {
        deep: !0
      }
    ), z(() => {
      x();
    });
    function x() {
      Number.isFinite(l.value) ? (l.future ? l.value > Date.now() ? r.value = l.value : H() : l.value > 0 ? r.value = l.value + Date.now() : H(), a.value = r.value - Date.now(), (l.future || !l.future && l.active) && (o.value && cancelAnimationFrame(o.value), o.value = requestAnimationFrame(g))) : a.value = 0;
    }
    function H() {
      a.value = 0, $("finish");
    }
    function g() {
      r.value > Date.now() ? (a.value = r.value - Date.now(), o.value = requestAnimationFrame(g)) : H();
    }
    function i(e, t = 2) {
      return String(e).padStart(t, "0");
    }
    function V(e) {
      let t = l.format;
      if (n.value.showMillisecond) {
        const w = e % 1e3;
        t = t.replace("SSS", i(w, 3));
      }
      e = Math.floor(e / 1e3);
      let f;
      n.value.showYear ? (f = Math.floor(e / (3600 * 24 * 30 * 12)), t = t.includes("YY") ? t.replace("YY", i(f)) : t.replace("Y", String(f))) : f = 0;
      let c;
      n.value.showMonth ? (e = e - f * 60 * 60 * 24 * 30 * 12, c = Math.floor(e / (3600 * 24 * 30)), t = t.includes("MM") ? t.replace("MM", i(c)) : t.replace("M", String(c))) : c = 0;
      let v;
      n.value.showDay ? (e = e - c * 60 * 60 * 24 * 30, v = Math.floor(e / (3600 * 24)), t = t.includes("DD") ? t.replace("DD", i(v)) : t.replace("D", String(v))) : v = 0;
      let d;
      n.value.showHour ? (e = e - v * 60 * 60 * 24, d = Math.floor(e / 3600), t = t.includes("HH") ? t.replace("HH", i(d)) : t.replace("H", String(d))) : d = 0;
      let p;
      if (n.value.showMinute ? (e = e - d * 60 * 60, p = Math.floor(e / 60), t = t.includes("mm") ? t.replace("mm", i(p)) : t.replace("m", String(p))) : p = 0, n.value.showSecond) {
        const w = e - p * 60;
        t = t.includes("ss") ? t.replace("ss", i(w)) : t.replace("s", String(w));
      }
      return t;
    }
    function _() {
      x();
    }
    return E({
      reset: _
    }), (e, t) => (s(), u("div", Z, [
      q.value ? (s(), u("div", {
        key: 0,
        class: "countdown-title",
        style: F(e.titleStyle)
      }, [
        y(e.$slots, "title", {}, () => [
          M(h(l.title), 1)
        ], !0)
      ], 4)) : m("", !0),
      I("div", j, [
        Y.value ? (s(), u(A, { key: 0 }, [
          Y.value || a.value > 0 ? (s(), u("span", G, [
            y(e.$slots, "prefix", {}, () => [
              M(h(e.prefix), 1)
            ], !0)
          ])) : m("", !0)
        ], 64)) : m("", !0),
        e.finish && a.value === 0 ? (s(), u("span", {
          key: 1,
          class: "time-value",
          style: F(e.valueStyle)
        }, [
          y(e.$slots, "finish", {}, () => [
            M(h(e.finish), 1)
          ], !0)
        ], 4)) : (s(), u("span", {
          key: 2,
          class: "time-value",
          style: F(e.valueStyle)
        }, h(V(a.value)), 5)),
        C.value ? (s(), u(A, { key: 3 }, [
          C.value || a.value > 0 ? (s(), u("span", J, [
            y(e.$slots, "suffix", {}, () => [
              M(h(e.suffix), 1)
            ], !0)
          ])) : m("", !0)
        ], 64)) : m("", !0)
      ])
    ]));
  }
});
export {
  O as default
};
