// vue-amazing-ui/es/index.js
/* empty css                  */
import * as o from "./components.js";
import { add as u, cancelRaf as d, dateFormat as p, debounce as m, downloadFile as x, formatNumber as i, rafTimeout as n, throttle as c, toggleDark as S, useEventListener as g, useFps as b, useInject as C, useMediaQuery as T, useMutationObserver as v, useOptionsSupported as I, useResizeObserver as R, useScroll as h, useSlotsExist as k } from "./utils/index.js";
import { VueAmazingUIResolver as P } from "./utils/resolver.js";
import { default as B } from "./alert/index.js";
import { default as F } from "./avatar/index.js";
import { default as L } from "./backtop/index.js";
import { default as N } from "./badge/index.js";
import { default as j } from "./breadcrumb/index.js";
import { default as z } from "./button/index.js";
import { default as U } from "./calendar/index.js";
import { default as W } from "./card/index.js";
import { default as H } from "./carousel/index.js";
import { default as J } from "./cascader/index.js";
import { default as X } from "./checkbox/index.js";
import { default as Z } from "./collapse/index.js";
import { default as $ } from "./colorpicker/index.js";
import { default as te } from "./configprovider/index.js";
import { default as re } from "./countdown/index.js";
import { default as fe } from "./datepicker/index.js";
import { default as le } from "./descriptions/descriptions/index.js";
import { default as de } from "./descriptions/descriptionsitem/index.js";
import { default as me } from "./dialog/index.js";
import { default as ie } from "./divider/index.js";
import { default as ce } from "./drawer/index.js";
import { default as ge } from "./ellipsis/index.js";
import { default as Ce } from "./empty/index.js";
import { default as ve } from "./flex/index.js";
import { default as Re } from "./floatbutton/index.js";
import { default as ke } from "./gradienttext/index.js";
import { default as Pe } from "./grid/row/index.js";
import { default as Be } from "./grid/col/index.js";
import { default as Fe } from "./highlight/index.js";
import { default as Le } from "./image/index.js";
import { default as Ne } from "./input/index.js";
import { default as je } from "./inputnumber/index.js";
import { default as ze } from "./inputsearch/index.js";
import { default as Ue } from "./list/list/index.js";
import { default as We } from "./list/listitem/index.js";
import { default as He } from "./loadingbar/index.js";
import { default as Je } from "./message/index.js";
import { default as Xe } from "./modal/index.js";
import { default as Ze } from "./notification/index.js";
import { default as $e } from "./numberanimation/index.js";
import { default as tt } from "./pagination/index.js";
import { default as rt } from "./popconfirm/index.js";
import { default as ft } from "./popover/index.js";
import { default as lt } from "./progress/index.js";
import { default as dt } from "./qrcode/index.js";
import { default as mt } from "./radio/index.js";
import { default as it } from "./rate/index.js";
import { default as ct } from "./result/index.js";
import { default as gt } from "./scrollbar/index.js";
import { default as Ct } from "./segmented/index.js";
import { default as vt } from "./select/index.js";
import { default as Rt } from "./skeleton/index.js";
import { default as kt } from "./slider/index.js";
import { default as Pt } from "./space/index.js";
import { default as Bt } from "./spin/index.js";
import { default as Ft } from "./statistic/index.js";
import { default as Lt } from "./steps/index.js";
import { default as Nt } from "./swiper/index.js";
import { default as jt } from "./switch/index.js";
import { default as zt } from "./table/index.js";
import { default as Ut } from "./tabs/index.js";
import { default as Wt } from "./tag/index.js";
import { default as Ht } from "./textarea/index.js";
import { default as Jt } from "./textscroll/index.js";
import { default as Xt } from "./timeline/index.js";
import { default as Zt } from "./tooltip/index.js";
import { default as $t } from "./upload/index.js";
import { default as to } from "./video/index.js";
import { default as ro } from "./waterfall/index.js";
import { default as fo } from "./watermark/index.js";
const r = function(e) {
  return Object.values(o).forEach((t) => {
    t.install && e.use(t);
  }), e;
}, f = {
  install: r
};
export {
  B as Alert,
  F as Avatar,
  L as BackTop,
  N as Badge,
  j as Breadcrumb,
  z as Button,
  U as Calendar,
  W as Card,
  H as Carousel,
  J as Cascader,
  X as Checkbox,
  Be as Col,
  Z as Collapse,
  $ as ColorPicker,
  te as ConfigProvider,
  re as Countdown,
  fe as DatePicker,
  le as Descriptions,
  de as DescriptionsItem,
  me as Dialog,
  ie as Divider,
  ce as Drawer,
  ge as Ellipsis,
  Ce as Empty,
  ve as Flex,
  Re as FloatButton,
  ke as GradientText,
  Fe as Highlight,
  Le as Image,
  Ne as Input,
  je as InputNumber,
  ze as InputSearch,
  Ue as List,
  We as ListItem,
  He as LoadingBar,
  Je as Message,
  Xe as Modal,
  Ze as Notification,
  $e as NumberAnimation,
  tt as Pagination,
  rt as Popconfirm,
  ft as Popover,
  lt as Progress,
  dt as QRCode,
  mt as Radio,
  it as Rate,
  ct as Result,
  Pe as Row,
  gt as Scrollbar,
  Ct as Segmented,
  vt as Select,
  Rt as Skeleton,
  kt as Slider,
  Pt as Space,
  Bt as Spin,
  Ft as Statistic,
  Lt as Steps,
  Nt as Swiper,
  jt as Switch,
  zt as Table,
  Ut as Tabs,
  Wt as Tag,
  Jt as TextScroll,
  Ht as Textarea,
  Xt as Timeline,
  Zt as Tooltip,
  $t as Upload,
  to as Video,
  P as VueAmazingUIResolver,
  ro as Waterfall,
  fo as Watermark,
  u as add,
  d as cancelRaf,
  p as dateFormat,
  m as debounce,
  f as default,
  x as downloadFile,
  i as formatNumber,
  r as install,
  n as rafTimeout,
  c as throttle,
  S as toggleDark,
  g as useEventListener,
  b as useFps,
  C as useInject,
  T as useMediaQuery,
  v as useMutationObserver,
  I as useOptionsSupported,
  R as useResizeObserver,
  h as useScroll,
  k as useSlotsExist
};
