// vue-amazing-ui/es/list/list/List.vue2.js
import { defineComponent as g, computed as d, createBlock as v, openBlock as t, unref as o, mergeProps as s, withCtx as y, createElementVNode as B, normalizeClass as k, createElementBlock as l, createCommentVNode as r, renderSlot as i, createTextVNode as p, toDisplayString as f, createVNode as u } from "vue";
import P from "../../spin/index.js";
import b from "../../empty/index.js";
import w from "../../pagination/index.js";
import { useSlotsExist as z } from "../../utils/index.js";
const E = {
  key: 0,
  class: "list-header"
}, C = {
  key: 2,
  class: "list-empty"
}, N = {
  key: 3,
  class: "list-footer"
}, S = {
  key: 4,
  class: "list-pagination"
}, T = /* @__PURE__ */ g({
  __name: "List",
  props: {
    bordered: { type: Boolean, default: !1 },
    vertical: { type: Boolean, default: !1 },
    split: { type: Boolean, default: !0 },
    size: { default: "middle" },
    loading: { type: Boolean, default: !1 },
    hoverable: { type: Boolean, default: !1 },
    header: { default: void 0 },
    footer: { default: void 0 },
    spinProps: { default: () => ({}) },
    emptyProps: { default: () => ({}) },
    showPagination: { type: Boolean, default: !1 },
    pagination: { default: () => ({}) }
  },
  setup(m) {
    const n = m, a = z(["header", "default", "footer"]), h = d(() => a.header || n.header), c = d(() => a.footer || n.footer);
    return (e, V) => (t(), v(o(P), s({
      size: "small",
      spinning: e.loading
    }, e.spinProps), {
      default: y(() => [
        B("div", {
          class: k(["list-wrap", {
            "list-bordered": e.bordered,
            "list-vertical": e.vertical,
            "list-split": e.split,
            "list-small": e.size === "small",
            "list-large": e.size === "large",
            "list-hoverable": e.hoverable
          }])
        }, [
          h.value ? (t(), l("div", E, [
            i(e.$slots, "header", {}, () => [
              p(f(e.header), 1)
            ], !0)
          ])) : r("", !0),
          o(a).default ? i(e.$slots, "default", { key: 1 }, void 0, !0) : (t(), l("div", C, [
            u(o(b), s({ image: "outlined" }, e.emptyProps), null, 16)
          ])),
          c.value ? (t(), l("div", N, [
            i(e.$slots, "footer", {}, () => [
              p(f(e.footer), 1)
            ], !0)
          ])) : r("", !0),
          e.showPagination ? (t(), l("div", S, [
            u(o(w), s({ placement: "right" }, e.pagination), null, 16)
          ])) : r("", !0)
        ], 2)
      ]),
      _: 3
    }, 16, ["spinning"]));
  }
});
export {
  T as default
};
