// vue-amazing-ui/es/slider/Slider.vue2.js
import { defineComponent as Ve, ref as k, computed as r, watch as ne, createElementBlock as m, openBlock as f, normalizeStyle as M, normalizeClass as R, unref as C, createElementVNode as L, createCommentVNode as I, Fragment as A, renderList as se, withModifiers as y, renderSlot as Be, createBlock as Re, createTextVNode as q, toDisplayString as G, resolveDynamicComponent as Ce, withKeys as S, isVNode as re } from "vue";
import { useInject as De, useResizeObserver as Ee } from "../utils/index.js";
const Te = { class: "slider-dots" }, je = { class: "slider-marks" }, Fe = ["onClick"], Ie = /* @__PURE__ */ Ve({
  __name: "Slider",
  props: {
    width: { default: "100%" },
    height: { default: "100%" },
    vertical: { type: Boolean, default: !1 },
    min: { default: 0 },
    max: { default: 100 },
    marks: { default: () => ({}) },
    disabled: { type: Boolean, default: !1 },
    range: { type: Boolean, default: !1 },
    step: { default: 1 },
    tooltip: { type: Boolean, default: !0 },
    tooltipOpen: { type: Boolean, default: !1 },
    tooltipStyle: { default: () => ({}) },
    formatTooltip: { type: Function, default: (Y) => Y },
    value: { default: 0 }
  },
  emits: ["update:value", "change"],
  setup(Y, { emit: ue }) {
    const l = Y, O = k(), Q = k(0), Z = k(0), n = k(0), a = k(0), w = k(), $ = k(), v = k(), d = k(), { colorPalettes: V } = De("Slider"), _ = ue, u = r(() => l.vertical ? Z.value : Q.value), ve = r(() => l.vertical ? {
      height: typeof l.height == "number" ? `${l.height}px` : l.height
    } : {
      width: typeof l.width == "number" ? `${l.width}px` : l.width
    }), fe = r(() => l.vertical ? {
      bottom: `${n.value}px`,
      top: "auto",
      height: `${a.value - n.value}px`
    } : {
      left: `${n.value}px`,
      right: "auto",
      width: `${a.value - n.value}px`
    }), de = r(() => l.vertical ? {
      bottom: `${n.value}px`,
      top: "auto",
      transform: "translate(-50%, 50%)"
    } : {
      left: `${n.value}px`,
      right: "auto",
      transform: "translate(-50%, -50%)"
    }), pe = r(() => l.vertical ? {
      bottom: `${a.value}px`,
      top: "auto",
      transform: "translate(-50%, 50%)"
    } : {
      left: `${a.value}px`,
      right: "auto",
      transform: "translate(-50%, -50%)"
    }), K = r(() => Object.keys(l.marks).length > 0), p = r(() => {
      let e = [];
      return K.value && (e = Object.keys(l.marks).map(parseFloat).sort((t, o) => t - o)), e;
    }), B = r(() => {
      const e = [];
      return p.value.length > 0 && p.value.forEach((t) => {
        e.push($e(t));
      }), e;
    }), h = r(() => l.step === "mark" ? 1 : l.step), ce = r(() => h.value.toString().split(".")[1]?.length ?? 0), ee = r(() => {
      let e = 0;
      return p.value.forEach((t) => {
        const o = t.toString().split(".");
        (o[1]?.length ?? 0 > e) && (e = o[1]?.length);
      }), e;
    }), P = r(() => {
      if (u.value === 0) return l.value;
      let e;
      if (a.value === u.value ? e = l.max : (e = le(a.value), l.step === "mark" ? e = H(e, p.value) : l.step > 1 && !p.value.includes(e) && (e = Math.round(e / l.step) * l.step)), l.range) {
        let t;
        return n.value === 0 ? t = l.min : (t = le(n.value), l.step === "mark" ? t = H(t, p.value) : l.step > 1 && !p.value.includes(t) && (t = Math.round(t / l.step) * l.step)), [t, e];
      }
      return e;
    }), me = r(() => l.range ? l.formatTooltip(P.value[0]) : null), he = r(() => l.range ? l.formatTooltip(P.value[1]) : l.formatTooltip(P.value));
    ne(
      () => [l.min, l.max, l.step, l.vertical, l.value],
      () => {
        te();
      },
      {
        deep: !0,
        flush: "post"
      }
    ), ne(P, (e) => {
      JSON.stringify(e) !== JSON.stringify(l.value) && (_("update:value", e), _("change", e));
    }), Ee(O, () => {
      ge();
    });
    function ge() {
      Q.value = O.value.offsetWidth, Z.value = O.value.offsetHeight, te();
    }
    function be(e) {
      const t = c((we(e) - l.min) / h.value, "*");
      n.value = g(t, 2);
    }
    function ke(e) {
      const t = c((Me(e) - l.min) / h.value, "*");
      a.value = g(t, 2);
    }
    function ye(e) {
      const t = c((Se(e) - l.min) / h.value, "*");
      a.value = g(t, 2);
    }
    function te() {
      l.range ? (be(l.value[0]), ke(l.value[1])) : ye(l.value);
    }
    function we(e) {
      return e < l.min ? l.min : e;
    }
    function Me(e) {
      return e > l.max ? l.max : e;
    }
    function Se(e) {
      return e < l.min ? l.min : e > l.max ? l.max : e;
    }
    function H(e, t) {
      if (!K.value)
        return console.warn("Please set the marks property"), 0;
      let o = t[0], i = Math.abs(e - o);
      const s = t.length;
      for (let N = 1; N < s; N++) {
        const ie = Math.abs(e - t[N]);
        ie < i && (i = ie, o = t[N]);
      }
      return o;
    }
    function le(e) {
      const t = c(e, "/") * h.value + l.min;
      return p.value.includes(g(t, ee.value)) ? g(t, ee.value) : g(t, ce.value);
    }
    function $e(e) {
      const t = (e - l.min) / (l.max - l.min) * u.value;
      return g(t, 2);
    }
    function g(e, t) {
      return parseFloat(e.toFixed(t));
    }
    function oe(e) {
      e.classList.remove("show-handle-tooltip");
    }
    function b(e, t) {
      e.focus(), l.tooltip && !l.tooltipOpen && t.classList.add("show-handle-tooltip");
    }
    function D(e) {
      let t, o;
      if (l.vertical) {
        t = O.value.getBoundingClientRect().bottom - e.clientY;
        const s = Math.round(c(t, "/"));
        o = g(c(s, "*"), 2);
      } else {
        const i = O.value.getBoundingClientRect().left;
        t = e.clientX - i;
        const s = Math.round(c(t, "/"));
        o = g(c(s, "*"), 2);
      }
      return {
        originalPosition: t,
        stepPosition: o
      };
    }
    function X(e, t) {
      if (K.value) {
        const o = H(e, B.value);
        return Math.abs(e - o) < Math.abs(e - t) ? o : t;
      } else
        return t;
    }
    function Pe(e) {
      const {
        originalPosition: t,
        // 鼠标点击位置
        stepPosition: o
        // 只考虑步长时将要移动的位置
      } = D(e);
      let i;
      if (l.step === "mark") {
        const s = H(t, B.value);
        l.range && Math.abs(s - n.value) < Math.abs(s - a.value) ? (s !== n.value && (n.value = s), b(w.value, $.value)) : (s !== a.value && (a.value = s), b(v.value, d.value));
      } else
        i = X(t, o), l.range && Math.abs(i - n.value) < Math.abs(i - a.value) ? (i !== n.value && (n.value = i), b(w.value, $.value)) : (i !== a.value && (a.value = i), b(v.value, d.value));
    }
    function J(e) {
      w.value && (document.addEventListener("mousemove", U), document.addEventListener("mouseup", E), U(e));
    }
    function U(e) {
      const {
        originalPosition: t,
        // 初始位置
        stepPosition: o
        // 只考虑步长时将要移动的位置
      } = D(e);
      let i;
      if (l.tooltip && !l.tooltipOpen && $.value.classList.add("show-handle-tooltip"), l.step === "mark") {
        const s = H(t, B.value);
        s <= a.value ? s !== n.value && (n.value = s) : (n.value = a.value, v.value.focus(), E(), W(e));
      } else
        i = X(t, o), i < 0 ? n.value = 0 : 0 <= i && i <= a.value ? n.value = i : (n.value = a.value, v.value.focus(), E(), W(e));
    }
    function E() {
      l.tooltip && !l.tooltipOpen && $.value.classList.remove("show-handle-tooltip"), document.removeEventListener("mousemove", U), document.removeEventListener("mouseup", E);
    }
    function W(e) {
      v.value && (document.addEventListener("mousemove", x), document.addEventListener("mouseup", T), x(e));
    }
    function x(e) {
      let {
        originalPosition: t,
        // 初始位置
        stepPosition: o
        // 只考虑步长时将要移动的位置
      } = D(e), i;
      if ({ originalPosition: t, stepPosition: o } = D(e), l.tooltip && !l.tooltipOpen && d.value.classList.add("show-handle-tooltip"), l.step === "mark") {
        const s = H(t, B.value);
        s >= n.value ? s !== a.value && (a.value = s) : (a.value = n.value, l.range && (w.value.focus(), T(), J(e)));
      } else
        i = X(t, o), i > u.value ? a.value = u.value : n.value <= i && i <= u.value ? a.value = i : (a.value = n.value, l.range && (w.value.focus(), T(), J(e)));
    }
    function T() {
      l.tooltip && !l.tooltipOpen && d.value.classList.remove("show-handle-tooltip"), document.removeEventListener("mousemove", x), document.removeEventListener("mouseup", T);
    }
    function Le(e) {
      const t = `${Math.abs(e - l.min) / (l.max - l.min) * 100}%`;
      return l.vertical ? {
        bottom: t
      } : {
        left: t
      };
    }
    function ae(e) {
      return l.range ? P.value[0] <= e && e <= P.value[1] : e <= P.value;
    }
    function j(e) {
      const t = l.marks[e], i = typeof t == "object" && !re(t) ? t.label : t;
      return i ? typeof i == "function" ? i() : i : null;
    }
    function Oe(e) {
      const t = `${Math.abs(e - l.min) / (l.max - l.min) * 100}%`;
      let o = {};
      const i = l.marks[e];
      return typeof i == "object" && !re(i) && "style" in i && (o = i.style), l.vertical ? {
        transform: "translateY(50%)",
        bottom: t,
        ...o
      } : {
        transform: "translateX(-50%)",
        left: t,
        ...o
      };
    }
    function He(e) {
      const t = B.value[e];
      l.range && Math.abs(t - n.value) < Math.abs(t - a.value) ? (t !== n.value && (n.value = t), b(w.value, $.value)) : (t !== a.value && (a.value = t), b(v.value, d.value));
    }
    function F(e, t) {
      const o = c(e, "-");
      t === "low" ? o < 0 ? n.value = 0 : n.value = o : o >= n.value ? a.value = o : (a.value = n.value, n.value = o, w.value.focus());
    }
    function z(e, t) {
      const o = c(e, "+");
      t === "high" ? o > u.value ? a.value = u.value : a.value = o : o <= a.value ? n.value = o : (n.value = a.value, a.value = o, v.value.focus());
    }
    function c(e, t) {
      return t === "+" ? e + u.value * h.value / (l.max - l.min) : t === "-" ? e - u.value * h.value / (l.max - l.min) : t === "*" ? e * u.value * h.value / (l.max - l.min) : t === "/" ? e * (l.max - l.min) / (u.value * h.value) : e;
    }
    return (e, t) => (f(), m("div", {
      ref_key: "sliderRef",
      ref: O,
      class: R(["slider-wrap", {
        "slider-horizontal": !e.vertical,
        "slider-vertical": e.vertical,
        "slider-with-marks": Object.keys(e.marks).length > 0,
        "slider-disabled": e.disabled
      }]),
      style: M([
        ve.value,
        `
        --slider-rail-color: rgba(0, 0, 0, 0.04);
        --slider-rail-color-hover: rgba(0, 0, 0, 0.1);
        --slider-rail-color-disabled: rgba(0, 0, 0, 0.06);
        --slider-track-color: ${C(V)[2]};
        --slider-track-color-hover: ${C(V)[5]};
        --slider-track-color-disabled: rgba(0, 0, 0, 0.25);
        --slider-handle-color: #fff;
        --slider-handle-shadow-color: ${C(V)[2]};
        --slider-handle-shadow-color-hover-focus: ${C(V)[5]};
        --slider-handle-shadow-color-disabled: #bfbfbf;
        --slider-dot-color: #fff;
        --slider-dot-border-color: #f0f0f0;
        --slider-dot-border-color-hover: rgba(0, 0, 0, 0.1);
        --slider-dot-color-active: ${C(V)[2]};
        --slider-mark-color: rgba(0, 0, 0, 0.45);
        --slider-mark-color-active: rgba(0, 0, 0, 0.88);
        --slider-tooltip-color: #fff;
        --slider-tooltip-bg-color: rgba(0, 0, 0, 0.85);
      `
      ]),
      onClick: t[12] || (t[12] = (o) => e.disabled ? () => !1 : Pe(o))
    }, [
      t[15] || (t[15] = L("div", { class: "slider-rail" }, null, -1)),
      L("div", {
        class: "slider-track",
        style: M(fe.value)
      }, null, 4),
      Object.keys(e.marks).length > 0 ? (f(), m(A, { key: 0 }, [
        L("div", Te, [
          (f(!0), m(A, null, se(p.value, (o, i) => (f(), m("span", {
            class: R(["slider-dot", { "slider-dot-active": ae(o) }]),
            style: M(Le(o)),
            key: i
          }, null, 6))), 128))
        ]),
        L("div", je, [
          (f(!0), m(A, null, se(p.value, (o, i) => (f(), m("span", {
            class: R(["slider-mark", { "slider-mark-active": ae(o) }]),
            style: M(Oe(o)),
            key: i,
            onClick: y((s) => e.disabled ? () => !1 : He(i), ["stop"])
          }, [
            Be(e.$slots, "mark", {
              label: j(o),
              value: o
            }, () => [
              typeof j(o) == "string" ? (f(), m(A, { key: 0 }, [
                q(G(j(o)), 1)
              ], 64)) : (f(), Re(Ce(j(o)), { key: 1 }))
            ], !0)
          ], 14, Fe))), 128))
        ])
      ], 64)) : I("", !0),
      e.range ? (f(), m("div", {
        key: 1,
        tabindex: "0",
        ref_key: "lowHandleRef",
        ref: w,
        class: "slider-handle",
        style: M(de.value),
        onKeydown: [
          t[0] || (t[0] = S(y((o) => e.disabled ? () => !1 : F(n.value, "low"), ["prevent"]), ["left"])),
          t[1] || (t[1] = S(y((o) => e.disabled ? () => !1 : z(n.value, "low"), ["prevent"]), ["right"])),
          t[2] || (t[2] = S(y((o) => e.disabled ? () => !1 : F(n.value, "low"), ["prevent"]), ["down"])),
          t[3] || (t[3] = S(y((o) => e.disabled ? () => !1 : z(n.value, "low"), ["prevent"]), ["up"]))
        ],
        onMousedown: t[4] || (t[4] = (o) => e.disabled ? () => !1 : J(o)),
        onBlur: t[5] || (t[5] = (o) => e.tooltip && !e.disabled && !e.tooltipOpen ? oe($.value) : () => !1)
      }, [
        e.tooltip ? (f(), m("div", {
          key: 0,
          ref_key: "lowTooltipRef",
          ref: $,
          class: R(["handle-tooltip", { "show-handle-tooltip": e.tooltipOpen }]),
          style: M(e.tooltipStyle)
        }, [
          q(G(me.value) + " ", 1),
          t[13] || (t[13] = L("div", { class: "tooltip-arrow" }, null, -1))
        ], 6)) : I("", !0)
      ], 36)) : I("", !0),
      L("div", {
        tabindex: "0",
        ref_key: "highHandleRef",
        ref: v,
        class: "slider-handle",
        style: M(pe.value),
        onKeydown: [
          t[6] || (t[6] = S(y((o) => e.disabled ? () => !1 : F(a.value, "high"), ["prevent"]), ["left"])),
          t[7] || (t[7] = S(y((o) => e.disabled ? () => !1 : z(a.value, "high"), ["prevent"]), ["right"])),
          t[8] || (t[8] = S(y((o) => e.disabled ? () => !1 : F(a.value, "high"), ["prevent"]), ["down"])),
          t[9] || (t[9] = S(y((o) => e.disabled ? () => !1 : z(a.value, "high"), ["prevent"]), ["up"]))
        ],
        onMousedown: t[10] || (t[10] = (o) => e.disabled ? () => !1 : W(o)),
        onBlur: t[11] || (t[11] = (o) => e.tooltip && !e.disabled && !e.tooltipOpen ? oe(d.value) : () => !1)
      }, [
        e.tooltip ? (f(), m("div", {
          key: 0,
          ref_key: "highTooltipRef",
          ref: d,
          class: R(["handle-tooltip", { "show-handle-tooltip": e.tooltipOpen }]),
          style: M(e.tooltipStyle)
        }, [
          q(G(he.value) + " ", 1),
          t[14] || (t[14] = L("div", { class: "tooltip-arrow" }, null, -1))
        ], 6)) : I("", !0)
      ], 36)
    ], 6));
  }
});
export {
  Ie as default
};
