// vue-amazing-ui/es/descriptions/descriptionsitem/DescriptionsItem.vue2.js
import { defineComponent as o, createElementBlock as d, openBlock as r, Fragment as i, createElementVNode as t, normalizeStyle as s, renderSlot as l, createTextVNode as a, toDisplayString as n } from "vue";
const p = ["data-span"], c = ["data-span"], b = /* @__PURE__ */ o({
  __name: "DescriptionsItem",
  props: {
    label: { default: void 0 },
    span: { default: void 0 },
    labelStyle: { default: () => ({}) },
    contentStyle: { default: () => ({}) }
  },
  setup(u) {
    return (e, m) => (r(), d(i, null, [
      t("div", {
        class: "descriptions-item",
        "data-span": e.span
      }, [
        t("span", {
          class: "descriptions-label",
          style: s(e.labelStyle)
        }, [
          l(e.$slots, "label", {}, () => [
            a(n(e.label), 1)
          ], !0)
        ], 4),
        t("span", {
          class: "descriptions-content",
          style: s(e.contentStyle)
        }, [
          l(e.$slots, "default", {}, void 0, !0)
        ], 4)
      ], 8, p),
      t("tr", {
        class: "descriptions-item-bordered",
        "data-span": e.span
      }, [
        t("th", {
          class: "descriptions-label-th",
          style: s(e.labelStyle)
        }, [
          l(e.$slots, "label", {}, () => [
            a(n(e.label), 1)
          ], !0)
        ], 4),
        t("td", {
          class: "descriptions-content-td",
          style: s(e.contentStyle)
        }, [
          l(e.$slots, "default", {}, void 0, !0)
        ], 4)
      ], 8, c)
    ], 64));
  }
});
export {
  b as default
};
